### PR TITLE
Add --seed(-s) parameter for reproducible problem generation

### PR DESCRIPTION
**Pull request by simonusher**
Hello!
We needed to have reproducible problem generation in our work so I added this parameter. 
If you feel it would be useful to have something like this then please review this.

I don't really have much experience with Rust programming so there could be mistakes. I also might have removed more than you wanted (codomain.rs line 131) to make the compiler happy. I pass a mutable rng there, so I couldn't use the parallel progress iterator.

### CODE REVIEW

File: src/problem/codomain.rs [hunk 1]
```diff
@@ -72,31 +73,40 @@ enum CodomainCommand {
 
 ///Run codomain generator from command line options (structopt)
 pub fn run_opt(codomain_opt: CodomainOpt) -> Result<(), Box<dyn Error>> {
-    match codomain_opt.codomain_command {
-        CodomainCommand::Folder { folder_paths } => {
-            for folder_path in folder_paths {
-                handle_folder(folder_path)?;
+    match codomain_opt {
+        CodomainOpt { codomain_command, seed } => {
```
Comment (tobiasvandriessel) [inline src/problem/codomain.rs]: You can probably move the multiple `let mut rng = get_rng(seed);` lines here, just after the `seed` is defined.

File: src/problem/problem_generation.rs [hunk 2]
```diff
@@ -90,48 +94,63 @@ enum ProblemCommand {
...
-        }
-        ProblemCommand::ConfigurationFolder {
-            folder_paths,
-            number_of_problems_to_generate,
-        } => {
-            for folder_path in folder_paths {
-                generate_codomain_and_problem_from_folder(
-                    &folder_path,
+    match problem_opt {
+        ProblemOpt { problem_command, seed } => {
```
Comment (tobiasvandriessel) [inline src/problem/problem_generation.rs]: Again, you can probably move the multiple let mut rng = get_rng(seed); lines here, just after the seed is defined.

File: src/problem/codomain.rs (hunk 1 above)
Comment (tobiasvandriessel) [reply src/problem/codomain.rs]: Also, I'd not do a match if there's no other options, although it _is_ more clear what properties the CodomainOpt struct has. So, I'd do it much like I did it before, except for a line `let mut rng = get_rng(codomain_opt.seed)` inserted right after the function signature declaration.

File: src/problem/problem_generation.rs (hunk 2 above)
Comment (tobiasvandriessel) [reply src/problem/problem_generation.rs]: Again, I'd not do a match if there's no other options, although it is more clear what properties the CodomainOpt struct has. So, I'd do it much like I did it before, except for a line let mut rng = get_rng(codomain_opt.seed) inserted right after the function signature declaration.

File: src/problem/codomain.rs [hunk 3]
```diff
@@ -118,8 +128,8 @@ fn handle_folder(folder_path: PathBuf) -> Result<(), Box<dyn Error>> {
         .collect();
 
     //And handle each of them
-    file_entries.into_par_iter().progress().for_each(|path| {
-        handle_input_configuration_file(path).unwrap();
+    file_entries.into_iter().for_each(|path| {
```
Comment (tobiasvandriessel) [inline src/problem/codomain.rs]: It's a shame the parallel iterator needs to go, but it's probably worth it given the added feature of reproducible randomness. I do wonder whether the `for_each` will always process the files in the same order and whether that matters then for the reproducibility.

Comment (tobiasvandriessel) [reply src/problem/codomain.rs]: And didn't the progress bar work anymore? I think I'd like to keep that actually.

**Review by tobiasvandriessel — CHANGES_REQUESTED** [review]
Thanks for creating a pull request for this; I actually spoke about this feature with somebody that's working on integrating this into a testing framework, so thank you for taking the time to make this work already :)

I do have some small comments, but nothing major. Would you be able to process my review comments? I don't mind doing it myself, but perhaps it's nice when you do it, since it's your pull request. Happy to help!

**Comment by simonusher** [comment]
Thanks for your reply. I addressed your comments, removed the unnecessary match statements, and moved the get_rng call up. I also restored the progress bar. I must have removed it by accident.
Next, taking what you said about file reading order into consideration, I checked the [documentation](https://doc.rust-lang.org/std/fs/fn.read_dir.html). It states that the order in which entries are returned from read_dir is not guaranteed,  so I added an explicit sorted() call, in places where you didn't sort the entries already. I'm convinced that it shouldn't break anything and ensure reproducible results, but I would really appreciate your review here. 
Lastly, the binary size also got a bit bigger, probably due to more imports. I hope that's not a problem.

**Comment by tobiasvandriessel** [comment]
Thanks @simonusher, I appreciate the effort you put in. Apologies for my late reply.
Really nice that you sorted the directory entries, that's very convenient! I changed my usage of sort() at some places to use .sorted() as well. Now I'll be testing whether the changes indeed do not break anything; for this I'll need to add some integration tests, so this might take me a bit of time.

**Comment by tobiasvandriessel** [comment]
@simonusher I've added a simple integration test, but I should definitely add more (unit & integration) tests later on. I consider the PR closed as the integration test works on 2 PCs and my LTGA experiment runner (private project) seems to be able to work nicely still, so the file IO is working as it should.

**Comment by tobiasvandriessel** [comment]
FYI: You can test whether the integration test is successful by running 'cargo test'.

**Comment by simonusher** [comment]
I'm sorry for not replying earlier. Thank you for incorporating this into your library. I hope it will find its uses.

**Comment by tobiasvandriessel** [comment]
No problem, and I am sure it will :)

**Pull request merged**